pulser-core/pulser/channels/comparison_tools.py

```diff
@@ -0,0 +1,201 @@
+
+    Attributes:
+        obj_dict: Dict of attributes and values of the object to compare.
+        best_obj_dict: Dict of attributes and values of the best object.
+        comparison_ops: Dict of attributes and comparison operators to
+            use to compare the object and the best object.
+
+    Returns:
+        True if the comparison works, raises a ValueError otherwise.
+    """
+    # If the two values are almost equal then there is no need to compare them
+    comparison_ops_keys = list(comparison_ops.keys())
+    for key in comparison_ops_keys:
+        if (
+            obj_dict[key] is not None
+            and best_obj_dict[key] is not None
+            and isinstance(obj_dict[key], numbers.Number)
+            and isinstance(best_obj_dict[key], numbers.Number)
+            and np.isclose(obj_dict[key], best_obj_dict[key], rtol=1e-14)
+        ):
+            comparison_ops.pop(key)
+    is_wrong_effective_ch = _compare_with_None(
+        comparison_ops, obj_dict, best_obj_dict
+    )
+    # Validates if no True in the dictionary of the comparisons
+    if not (True in is_wrong_effective_ch.values()):
+        return True
+    is_wrong_effective_index = list(is_wrong_effective_ch.values()).index(True)
+    is_wrong_key = list(is_wrong_effective_ch.keys())[is_wrong_effective_index]
+    raise ValueError(
+        f"{is_wrong_key} cannot be"
+        + (
+            " below "
+            if comparison_ops[is_wrong_key] == lt
+            else (
+                " above "
+                if comparison_ops[is_wrong_key] == gt
+                else " different than "
+            )
+        )
+        + f"{best_obj_dict[is_wrong_key]}."
+    )
+
+
+def validate_channel_from_best(
+    channel: Channel, best_channel: Channel
+) -> bool:
+    """Checks that a channel can be realized from another one.
+
+    Attributes:
+        channel: The channel to check.
+        best_channel: The channel that should have better properties.
+    """
+    if type(channel) != type(best_channel):
+        raise ValueError(
+            "Channels do not have the same types, "
+            f"{type(channel)} and {type(best_channel)}"
+        )
+    if channel.eom_config:
+        if best_channel.eom_config:
+            validate_eom_from_best(channel.eom_config, best_channel.eom_config)
+        else:
+            raise ValueError(
+                "eom_config cannot be defined in channel as the best_channel"
+                " does not have one."
+            )
+    best_ch_att = asdict(best_channel)
+    ch_att = asdict(channel)
+
+    # Error if attributes in channel and best_channel compare to True
+    comparison_ops = {
+        "addressing": ne,
+        "max_abs_detuning": gt,
+        "max_amp": gt,
+        "min_retarget_interval": lt,
+        "fixed_retarget_t": lt,
+        "max_targets": gt,
+        "clock_period": lt,
+        "min_duration": lt,
+        "max_duration": gt,
+        "mod_bandwidth": gt,
```

Comment:
I'm afraid things are not this simple for the quantities that affect the sequence when it is being reconstructed. For example, `mod_bandwidth` influences things like the wait times, so if you are rebuilding the sequence with a device that has a different modulation bandwidth (even if higher), it may turn out different.

I suggest you take a look at how `Sequence.switch_device()` works (particularly when `strict=True`) - where, by the way, I see these functions fitting very naturally

Reply:
I now see the mistake with my approach: it completely forgets the original goal of this validation, which is to check whether sequence from the the cloud can be performed with the available_devices. As a reminder:
```
        if emulator is None:
            available_devices = self.fetch_available_devices()
            # TODO: Could be better to check if the devices are
            # compatible, even if not exactly equal
            if sequence.device not in available_devices.values():
                raise ValueError(
                    "The device used in the sequence does not match any "
                    "of the devices currently available through the remote "
                    "connection."
                )
```
I only wanted to check if the values put in Device/Channel/EOM matched with Device/Channel/EOM, but that is not the topic indeed. The checks should garantee that the shape of the curve will be the same. So indeed the mod bandwidth should be the same. The clock_period should be the same as well (or a divider of the clock period of the best channel).

Here is a list of things that are different in `Sequence.switch_device`:
- takes into account whether the channels are reusable or not when matching channels of the devices.
- `basis` should be equal.
- if `strict`, the eom configs have to be exactly the same. To me, that's right for all the attributes except for controlled_beams: we can make a list of the controlled beams config ( RED, BLUE, (BLUE, RED) ) used in each eom block of the sequence, and check if they are among the controlled beams of the new channel (if the current device has controlled beams (RED, BLUE, (BLUE, RED)) but only the RED is controlled in the EOM blocks of the Sequence, a Channel with controlled beams (RED) would work `strictly`).
- if `strict`, `fixed_retarget_t` should match exactly. Given its definition "Time taken to change the target (in ns)." that's correct.
- if `strict`, `min_retarget_interval` should match exactly. As this is for testing only, and as its definition is "Minimum time required between the ends of two target instructions (in ns).", I think that can be changed into `new_channel.min_retarget_interval <= old_channel.min_retarget_interval`.
- tests all the other properties of device by implementing the sequence with it.
- Misses the comparison of layout/register of the Sequence with the calibrated_layouts of the new device (if there is a matching layout, the register of the sequence should have this new layout)

Reply:
To add another remark, the difference between `Sequence.switch_device` and the validation functions is that in `Sequence.switch_device`, the default operator is `ne`, whereas in the validation functions they are `None`, meaning that no comparisons can be done. If I were to keep these validation functions, I would build the dictionary of comparison operators by first assigning to all the attributes of device/channel/eom `ne`, and then specify a different operator for some of them.

Reply:
> * if `strict`, the eom configs have to be exactly the same. To me, that's right for all the attributes except for controlled_beams: we can make a list of the controlled beams config ( RED, BLUE, (BLUE, RED) ) used in each eom block of the sequence, and check if they are among the controlled beams of the new channel (if the current device has controlled beams (RED, BLUE, (BLUE, RED)) but only the RED is controlled in the EOM blocks of the Sequence, a Channel with controlled beams (RED) would work `strictly`).

I agree with this, the problem is checking it... The decision of which beams are controlled depends on the value of `detuning_on`, which may even be parametrized, so we can't always know in advance. And even when we can, we currently don't store the controlled beams in in each block so we have to reverse engineer it, which is a pain...